File: fbi_web/src/com/fr/bi/web/js/data/filter/condition/general.or.filter.js
;
!(function(){
    BI.GeneralOrFilter = function(filter){
        if(BI.has(filter, "filter_value")){
            var filter_value = filter.filter_value || [];
            this.childs = BI.map(filter_value, function(idx, fv){
                return BI.FilterFactory.parseFilter(fv);
            })
        }
    };
    BI.GeneralOrFilter.prototype = {
        constructor: BI.GeneralOrFilter,

        isQualified: function(value, array){
            var res = BI.find(this.childs, function(idx, child){
                if(BI.isNotNull(child) && !child.isQualified(value, array)){
                    return true;
                }
            });
            return BI.isNull(res);
        },

        getFilterResult: function(array) {
            var self = this;
            return BI.filter(array, function(idx, val){
                return self.isQualified(val, array);
            });
        }
    }
})();

File: fbi_engine_analyse/src/com/fr/bi/cal/analyze/report/report/widget/chart/filter/condition/GeneralOrFilter.java
package com.fr.bi.cal.analyze.report.report.widget.chart.filter.condition;

import com.fr.bi.cal.analyze.report.report.widget.chart.filter.FilterFactory;
import com.fr.bi.cal.analyze.report.report.widget.chart.filter.IFilter;
import com.fr.json.JSONArray;
import com.fr.json.JSONException;
import com.fr.json.JSONObject;

import java.util.HashSet;
import java.util.Set;

/**
 * or condition
 * Created by AstronautOO7 on 2016/9/19.
 */
public class GeneralOrFilter implements IFilter {
    private IFilter[] childs;

    public GeneralOrFilter(JSONObject jo) throws JSONException {
        JSONArray filterValue;
        if (jo.has("filterValue")) {
            filterValue = jo.getJSONArray("filterValue");
            childs = new IFilter[filterValue.length()];
            for (int i = 0; i < filterValue.length(); i++) {
                childs[i] = FilterFactory.parseFilter(filterValue.getJSONObject(i));
            }
        }
    }

    @Override
    public JSONArray getFilterResult(JSONArray array) throws JSONException {
        Set<String> result = new HashSet<String>();
        for (int i = 0; i < childs.length; i++) {
            JSONArray ja = childs[i].getFilterResult(array);
            for (int k = 0; k < ja.length(); k++) {
                String key = ja.getString(k);
                result.add(key);
            }
        }
        return new JSONArray(result.toString());
    }
}
